File: CS 201 (Algorithm)/Lecture Qns/Python/Job Scheduling.py
# Single Server
class Solution1:
    def minimize_time_in_system(self, jobs):
        jobs.sort(key=lambda x: x[1])

        total_time = 0
        current_time = 0

        for job, processing_time in jobs:
            current_time += processing_time
            total_time += current_time

        return total_time


jobs = [('j1', 3), ('j2', 1), ('j3', 2), ('j4', 5)]
sol1 = Solution1()
print(sol1.minimize_time_in_system(jobs))

#On multiple servers

import heapq
class Solution2:
    def minimize_time_on_m_servers(self, jobs, m):
        jobs.sort(key=lambda x: x[1])
        servers = [0] * m
        total_time = 0

        for job, processing_time in jobs:
            earliest_server = heapq.heappop(servers)
            total_time += earliest_server + processing_time
            heapq.heappush(servers, earliest_server + processing_time)

        return total_time

jobs = [('j1', 3), ('j2', 1), ('j3', 2), ('j4', 5)]
sol2 = Solution2()
print(sol2.minimize_time_on_m_servers(jobs, 2))

File: CS 201 (Algorithm)/Lecture Qns/c++/Job_scheduling.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution1
{
    public:
        int min_time_single_sys(vector<pair<string,int>> &jobs)
        {
            sort(jobs.begin(),jobs.end(),[](pair<string,int> &a, pair<string,int> &b)
            {
                return a.second <= b.second;
            });

            int tot_time = 0;
            int curr_time = 0;

            for(auto job: jobs)
            {
                curr_time += job.second;
                tot_time += curr_time;
            }

            return tot_time;
        }
};


class Solution2
{
    public:
        int min_time_multiple_sys(vector<pair<string,int>> &jobs, int m)
        {
            sort(jobs.begin(),jobs.end(),[](pair<string,int> &a, pair<string,int> &b)
            {
                return a.second <= b.second;
            });

            priority_queue<int,vector<int>,greater<int>> servers;

            for (int i=0;i<m;i++)
                servers.push(0);

            int total_time = 0;
            for(auto job: jobs)
            {
                int earliest_server = servers.top();
                servers.pop(); // return type is void unlike that in python

                int new_time = earliest_server + job.second;
                servers.push(new_time);
                total_time += new_time;
            }

            return total_time;
        }
};

int main()
{
    vector<pair<string, int>> jobs = {{"j1", 3}, {"j2", 1}, {"j3", 2}, {"j4", 5}};
    Solution2 sol2;
    cout << sol2.min_time_multiple_sys(jobs, 2) << endl;

    Solution1 sol1;
    cout << sol1.min_time_single_sys(jobs) << endl;
    return 0;
}